File: code/python/data_structures_article-15.py
"""Queue implementation stack and stack implementation queue.

We keep one real queue of numbers and use one helper move to make `pop()` feel like stack behaviour.
"""

from __future__ import annotations

from queue import Queue
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Build and return a `MyStack` toy object for hands-on practice."""
    return MyStack()


class MyStack:
    def __init__(self):
        self.q = Queue()
        self.top_elem = 0

    def push(self, x: int) -> None:
        self.q.put(x)
        self.top_elem = x

    def pop(self) -> int:
        size = self.q.qsize()
        while size > 2:
            self.q.put(self.q.get())
            size -= 1
        self.top_elem = self.q.queue[0]
        self.q.put(self.q.get())
        return self.q.get()

    def top(self) -> int:
        return self.top_elem

    def empty(self) -> bool:
        return self.q.empty()

File: code/javascript/data_structures_article-15.js
/**
 * Queue implementation stack and stack implementation queue
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return MyStack(...args);
}

var MyStack = function() {
    this.q = [];
    this.top_elem = 0;
};
MyStack.prototype.push = function(x) {
    this.q.push(x);
    this.top_elem = x;
};
MyStack.prototype.top = function() {
    return this.top_elem;
};
MyStack.prototype.pop = function() {
    var size = this.q.length;
    while (size > 2) {
        this.q.push(this.q.shift());
        size--;
    }
    this.top_elem = this.q[0];
    this.q.push(this.q.shift());
    return this.q.shift();
};
MyStack.prototype.empty = function() {
    return this.q.length === 0;
};
